examples/spells/BoundedUInt.qnt

```diff
@@ -0,0 +1,607 @@
+module BoundedUInt {
+  /// The size of this integer type in bits.
+  const BITS: int
+
+  /// The smallest value that can be represented by this integer type.
+  pure val MIN = 0
+
+  /// The largest value that can be represented by this integer type.
+  pure val MAX = (2^BITS) - 1
+  
+  /// Record-representation of an unsigned bounded integer.
+  /// If the `error` field is nonempty, the record represents an exception of some
+  /// sort happening during computation, such as an overflow.
+  /// Otherwise, the record represents the integer `v`, such that `MIN <= v <= MAX`. 
+  type T = { v: int, error: str }
+
+  /// Constructs a bounded unsigned integer from a mathematical integer. 
+  /// If the mathematical integer lies outside the `[MIN, MAX]` interval, 
+  /// the error field will be nonempty.
+  pure def UInt(x: int): T = {v: x, 
+    error: 
+      if (MIN <= x and x <= MAX) ""
+      else "out of range"}
+
+  /// Given a record representation of a bounded unsigned integer, returns
+  /// true iff the record value lies in the `[MIN, MAX]` interval.
+  pure def isInRange(x: T): bool = and {
+    x.v >= MIN,
+    x.v <= MAX
+  }
+
+  /// Given a record representation of a bounded unsigned integer, returns
+  /// true iff the record value lies in the `[MIN, MAX]` interval, and the error field is empty.
+  pure def isValid(x: T): bool = and {
+    x.error == "",
+    isInRange(x)
+  }
+
+  /// Given a (T,T) => T operation `op`, which assumes its inputs are valid (w.r.t. `isValid`),
+  /// and two T-typed values, l and r, returns op(l,r) iff l and r are both valid. Otheriwse,
+  /// it returns some arbitrary invalid value.
+  pure def wrapErrorBin(lhs: T, rhs: T, op: (T, T) => T): T =
+    if (not(isValid(lhs))) lhs
+    else if (not(isValid(rhs))) rhs
+    else op(lhs, rhs)
+
+  /// Computes the absolute difference between lhs and rhs.
+  /// Assumes lhs and rhs are valid (w.r.t. `isValid`).
+  pure def absDiffUnsafe(lhs: T, rhs: T): T = {
+    val res = lhs.v - rhs.v
+    { 
+      v: if (res < 0) -res else res, 
+      error: "" 
+    }
+  }
+
+  /// Computes the absolute difference between lhs and rhs.
+  pure def absDiff(lhs: T, rhs: T): T = wrapErrorBin(lhs, rhs, absDiffUnsafe)
+  
+    ////////////////////////
+   // CHECKED OPERATIONS //
+  ////////////////////////
+
+  /// Unsafe checked integer addition. 
+  /// Computes `lhs + rhs`, setting the error field to "overflow" if overflow occurred.
+  /// Assumes lhs and rhs are valid (w.r.t. `isValid`).
+  pure def checkedAddUnsafe(lhs: T, rhs: T): T = {
+    val res = lhs.v + rhs.v
+    { 
+      v: res,
+      error: if (res > MAX) "overflow" else ""
+    }
+  }
```

Comment:
All of the "unsafe" methods are named as such, because they perform no validation on their inputs, i.e. they do not check whether `lhs` and `rhs` are valid representations of `UIntX`.

Reply:
For example,
```
>> val a = UInt(-10) // {v: -10, error: "out of range"}
>> val b = UInt(0)
>> checkedAddUnsafe(a,b)
{ v: -10, error: "" } // <-- It is not the case that (-10) > MAX, so error = ""
```

Reply:
How about getting rid of the unsafe/safe operators, and just call `isValid` at the end of every operator application? Then everything is safe, we can cut the code in about half, and we need to perform fewer checks. WDYT?

Reply:
There's a few problems:
 1) This basically forces us to go with an `error: bool` approach
 2) This misses cases, such as
```
>> val a = UInt(-10) // {v: -10, error: "out of range"}
>> val b = UInt(10)
>> checkedAddUnsafe(a,b)
{ v: 0, error: "" } // bad input, but magically fixed by math
```

Reply:
Ah I see.  Math is magic :D